api/src/controllers/login.js

```diff
@@ -300,26 +341,85 @@ const renderLogin = (req) => {
   return render('login', req);
 };
 
+const renderPasswordReset = (req) => {
+  return render('passwordReset', req);
+};
+
+const validatePasswordReset = (password, confirmPassword) => {
+  const error = validatePassword(password, confirmPassword);
+
+  if (!error) {
+    return { isValid: true };
+  }
+
+  return {
+    isValid: false,
+    error: ERROR_KEY_MAPPING[error.message.translationKey],
+    params: error.message.translationParams
+  };
+};
+
+const validateSession = async (req) => {
+  const sessionRes = await createSession(req);
+  if (sessionRes.statusCode !== 200) {
+    const error = new Error('Not logged in');
+    error.status = sessionRes.statusCode;
+    error.error = 'Not logged in';
+    throw error;
+  }
+  return sessionRes;
+};
+
+const sendLoginErrorResponse = (e, res) => {
+  if (e.status === 401) {
+    return res.status(401).json({ error: e.error });
+  }
+  logger.error('Error logging in: %o', e);
+  return res.status(500).json({ error: 'Unexpected error logging in' });
+};
+
 const login = async (req, res) => {
   try {
-    const sessionRes = await createSession(req);
-    if (sessionRes.statusCode !== 200) {
-      res.status(sessionRes.statusCode).json({ error: 'Not logged in' });
-    } else {
-      const redirectUrl = await setCookies(req, res, sessionRes);
-      res.status(302).send(redirectUrl);
-    }
+    const sessionRes = await validateSession(req);
+    const redirectUrl = await setCookies(req, res, sessionRes);
+    res.status(302).send(redirectUrl);
   } catch (e) {
-    if (e.status === 401) {
-      return res.status(401).json({ error: e.error });
-    }
-    logger.error('Error logging in: %o', e);
-    res.status(500).json({ error: 'Unexpected error logging in' });
+    return sendLoginErrorResponse(e, res);
   }
 };
 
+const updatePassword = async (user, newPassword) => {
+  const updatedUser = {
+    ...user,
+    password: newPassword,
+    password_change_required: false
+  };
+  await db.users.put(updatedUser);
+  // creating new session immediately after changing a password might 401
+  await new Promise(resolve => setTimeout(resolve, 50));
```

Comment:
This isn't a great solution because while it might work on your machine, what happens if couchdb is over a laggy network, or under load? A better solution would be to retry creating the session with 50ms delay until it works or some limit is reached. You'll have to be careful of rate limiting here...